content/courses/Ada_For_The_C_Embedded_Developer/chapters/05_SPARK.rst

```diff
@@ -365,260 +552,277 @@ SPARK checks with the :ada:`SPARK_Mode` aspect:
+
+   procedure Compute (K : Float; Z : out Integer; Flag : out Boolean) is
+      X : constant Float := Sin (K);
+   begin
+      if X < 0.0 then
+         Z := 0;
+         Flag := True;
+      elsif X > 0.0 then
+         Z := 1;
+         Flag := True;
+      else
+         Flag := False;
+      end if;
+   end Compute;
+
+:program:`gnatprove` tells us that :ada:`Z` might not be initialized 
+(assigned a value) in :ada:`Compute`, and indeed that is correct. 
+:ada:`Z` is a mode :ada:`out` parameter so the routine should assign a 
+value to it: :ada:`Z` is an output, after all. The fact that 
+:ada:`Compute` does not do so is a significant and nasty bug. Why is it 
+so nasty? In this case, formal parameter :ada:`Z` is of the scalar type 
+:ada:`Integer`, and scalar parameters are always passed by copy in Ada 
+and SPARK. That means that, when returning to the caller, an integer 
+value is copied to the caller's argument passed to :ada:`Z`. But this 
+procedure doesn't always assign the value to be copied back, and in that 
+case an arbitrary value |mdash| whatever is on the stack |mdash| is 
+copied to the caller's argument. The poor programmer must debug the code 
+to find the problem, yet the effect could appear well downstream from 
+the call to :ada:`Compute`. That's not only painful, it is expensive. 
+Better to find the problem before we even compile the code. 
+
+
+Contract-Based Programming
+--------------------------
 
-These contracts can be built after the values of function, parameters, global
-data or references to the current instance. For example:
+So far, we've seen assertions in a routine's sequence of statements, 
+either through implicit language-defined checks (is the index in the 
+right range?) or explicit user-defined checks. These checks are already 
+useful by themselves but they have an important limitation: the assertions 
+are in the implementation, hidden from the callers of the routine. For 
+example, a call's success or failure may depend upon certain input 
+values but the caller doesn't have that information. 
+
+Generally speaking, Ada and SPARK put a lot of emphasis on strong,
+complete specifications for the sake of abstraction and analysis.
+Callers need not examine the implementations to determine
+whether the arguments passed to it are changed, for example. It is
+possible to go beyond that, however, to specify arbitrary implementation
+constraints and behavior requirements via contracts.
+
+At the language level, contracts are higher-level forms of assertions
+associated with specifications and declarations, rather than sequences
+of statements. Like other assertions they can be activated or
+deactivated at run-time, and can be statically proven. We'll concentrate
+here on two kinds of contracts, both associated especially (but not
+exclusively) with procedures and functions:
+
+- *Preconditions*, those boolean conditions expected to be true *prior* to the call of a given subprogram 
```

Comment:
Editorial remark: I usually write "Boolean" instead of "boolean" because it refers to George Boole.